### PR TITLE
Add tests for frontend.

### PR DESCRIPTION
**Pull request by meandavejustice**
Tests will be written in [tape](https://github.com/substack/tape)
They are run with locally with [testling-cli](https://github.com/substack/testling)
They are piped to [faucet](https://github.com/substack/faucet) for pretty output.

todo:
- [x] Get remaining tests to pass (waiting on a change in #371)
- [x] set up coverage stats (coverify maybe)
- [x] add documentation for testing
- [x] add option to run tests in browser(they are currently running headless only)
- [x] hook up to run on travis.

The list of tests are [here](https://github.com/mozilla/idea-town/issues/147#issuecomment-168237808)

cc/ @lmorchard @chuckharmston @6a68 for discussion.

### CODE REVIEW

**Comment by lmorchard** [comment]
This is kind of a general process comment / question, but: Would it be better to keep work-in-progress things like this out of the PR queue? Instead we can just work from issues & branches, invite look-sees & discussion there. 

FWIW, this isn't unique to this project - it's come up as a small debate on pretty much every project I've been on :) 

I cruise the PR queue for things I can jump on and shepherd into master branch. Unmergeable WIP PRs like this just stick there as noise. Occasionally I've accidentally merged WIP PRs and some of mine have been merged too.

**Comment by meandavejustice** [comment]
@lmorchard I've never really liked WIP prs either, just gotten in the habit from other teams i've worked on. I'm :+1: for only tracking via branches/issues from now on. Would you be okay with starting _after_ this one though? I should have this one finished up today.

**Comment by lmorchard** [comment]
Per the status meeting, you can ignore my grumblings for sure on this PR - and possibly on future ones too :)

File: idea_town/frontend/static-src/test/general.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+var test = require('tape');
+var path = require('path');
+var fs = require('fs');
+
+test(`🖳  Running Tests for ${__filename} ␄`, function(a) {
```
Comment (lmorchard) [inline idea_town/frontend/static-src/test/general.js]: Hmm, node v0.12 on Travis seems to dislike some characters in this line.

Node on my machine dislikes them too (v0.12.7)

Oh, actually I think it might be an issue with the ES6 template string... 0.12 doesn't support them. Are you running these tests in one of the Docker containers? If not you may have a different version of node than me & travis.

v0.12 isn't exactly a new version of node, so we can look into upgrading in Docker & on Travis... but that might be a task out of scope for this particular issue

Comment (meandavejustice) [inline idea_town/frontend/static-src/test/general.js]: I'm removing them for this test. Since this test is run outside of the babel transform.

**Comment by meandavejustice** [comment]
@lmorchard @6a68 r?

**Comment by lmorchard** [comment]
Taking a look... 

[Travis seems to generally like the tests](https://travis-ci.org/mozilla/idea-town/jobs/102016897#L702), but not sure what's up with the [unknown unicode character](https://github.com/mozilla/idea-town/pull/354/files#diff-d2dbb9957d3b230f8edea716289c5fc3R6) before each "Running tests for" line.

File: idea_town/frontend/static-src/app/views/#experiment-page.js# [hunk 2]
```diff
@@ -0,0 +1,249 @@
+import app from 'ampersand-app';
```
Comment (lmorchard) [inline idea_town/frontend/static-src/app/views/#experiment-page.js#]: Is this file meant to be here? `#experiment-page.js#` seems like maybe a backup filename or something.

File: TESTING.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Testing
+
+We currently have two different ways to write tests. Server tests are fairly
+standard [django tests](https://docs.djangoproject.com/en/1.9/topics/testing/).
+The client tests are written with [tape](https://www.npmjs.com/package/tape), and
+run with [testling](https://www.npmjs.com/package/testling).
+
+Here are a couple of tips to helpout when testing clientside code:
```
Comment (lmorchard) [inline TESTING.md]: Nits:
`s/helpout/help out/`
`s/clientside/client-side/`

File: idea_town/frontend/static-src/app/views/#experiment-page.js# (hunk 2 above)
Comment (meandavejustice) [inline idea_town/frontend/static-src/app/views/#experiment-page.js#]: whoops, nice catch. Need to add all of these to my global gitignore.

**Comment by lmorchard** [comment]
Some docs on running these tests locally would be handy. I've tried `npm test` as it appears in `tox.ini`, but it errors out with "No headless browser found". Tried `gulp test`, but it just seems to sit there. Not sure what it's for.

**Comment by lmorchard** [comment]
This is a good & massive PR. Would be really worthwhile to amend & expand the commit message. Add some details that aren't obviously explained by "add tests for frontend". For example:
- Added dependencies to package.json 
  - (what do they all do? maybe copy some of the comments from the issue into the commit message?)
- Moved external templates inline with views 
  - (_might_ be worth some git surgery to split those changes off into a separate commit to de-tangle the changes. But, that might also be more trouble than it's worth.)

IMO, more details the better in the commit message - especially for large PRs. It comes in handy later when someone uses `git blame` to figure out when & why something was changed.

**Comment by lmorchard** [comment]
Hmm, looks like if I run `gulp test` and then visit http://192.168.99.1:9966/, I see test results. Is there any way to run these in a console? I want to hook these up to a file watcher so they'll auto-run when I save changes. Not sure if it's a helpful suggestion at this point, but I've had success with [Karma](https://karma-runner.github.io/0.13/index.html) - it uses websockets to drive a browser & reports in the console.

**Comment by meandavejustice** [comment]
@lmorchard `gulp test` runs `npm run test:browser:live` which watches the `test/` dir and reloads the tests in the browser. These cannot be run in the console, but the simple `npm test` or `npm test:client` commands can.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -70,7 +77,7 @@ Reviewers will run through roughly the following checklist:
 - Are boundary cases considered?
 - Are error cases handled?
 
-(Once we have some tests, the reviewer will check that all the tests are passing,
+(the reviewer will check that all the tests are passing,
```
Comment (jaredhirsch) [inline CONTRIBUTING.md]: maybe just pull this into the list of bullet points above. speaking of which, does that list of reviewer things to look for actually track with how reviews are done? if not, should we modify the list or just remember to use the list when reviewing? cc @lmorchard

File: gulpfile.js [hunk 5]
```diff
@@ -100,6 +101,8 @@ gulp.task('scripts', ['lint'], function scriptsTask() {
   return bundledStream;
 });
 
+gulp.task('test', shell.task('npm run test:browser:live'));
```
Comment (jaredhirsch) [inline gulpfile.js]: It seems funny to use both gulp and npm for running the same set of tasks. Is there any reason we want it in both places, or can we standardize on one?

File: idea_town/frontend/static-src/test/lib/router.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  app.on('router:new-page', evt => {
+    if (evt.page === 'landing') {
+      t.ok(evt.page, 'Should redirect to landing if missing addon or creds');
+        app.me.hasAddon = true;
+        app.router.landing()
+      }
+
+    if (evt.page === 'home') {
+      t.ok(evt.page, 'Should redirect to home if logged in with creds');
+      app.reset();
```
Comment (jaredhirsch) [inline idea_town/frontend/static-src/test/lib/router.js]: If `app.reset` needs to be called after every test completes, you'll need to make sure it gets called even if the test fails or throws. For this reason, it's a good idea to extract it out to a test runner.

File: idea_town/frontend/static-src/test/lib/router.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
+import test from 'tape';
+import Router from '../../app/lib/router';
+import app from 'ampersand-app';
+import Me from '../../app/models/me';
+import Experiments from '../../app/collections/experiments';
+
+test(`🖳  Running Tests for ${__filename} ␄`, a => a.end());
```
Comment (jaredhirsch) [inline idea_town/frontend/static-src/test/lib/router.js]: Seems like this bit of boilerplate could be extracted out to a runner or parent class, too.

File: idea_town/frontend/static-src/test/lib/router.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+test('landing redirects correctly if logged in with addon', t => {
+  t.plan(2);
+
+  app.me = new Me({
+    user: {
+      id: 'gary@busey.biz',
+        addon: {url: '/lol nothing matters'}
+      }
+  });
```
Comment (jaredhirsch) [inline idea_town/frontend/static-src/test/lib/router.js]: Looks like quite a bit of duplicated test setup in this file. Can you extract this to a setup method?

File: idea_town/frontend/static-src/test/views/contributor-view.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import test from 'tape';
+import view from '../../app/views/contributor-view';
+
+test(`🖳  Running Tests for ${__filename} ␄`, a => a.end());
+
+test('Contibutor view renders', t => {
```
Comment (jaredhirsch) [inline idea_town/frontend/static-src/test/views/contributor-view.js]: 'Contributor'

File: idea_town/experiments/models.py [hunk 10]
```diff
@@ -42,6 +43,11 @@ class Meta:
     created = models.DateTimeField(auto_now_add=True)
     modified = models.DateTimeField(auto_now=True)
 
+    @cached_property
+    def installation_count(self):
+        return UserInstallation.objects.distinct('user').filter(
+            experiment=self).count()
+
```
Comment (jaredhirsch) [inline idea_town/experiments/models.py:50]: Hmm, is this supposed to be here? Seems like a rebase might be needed?

File: CONTRIBUTING.md (hunk 4 above)
Comment (meandavejustice) [inline CONTRIBUTING.md]: I guess _travis_ checks to see that the tests are run.

File: idea_town/experiments/models.py (hunk 10 above)
Comment (meandavejustice) [inline idea_town/experiments/models.py:50]: woah nice catch. lemme get rid of this

File: gulpfile.js (hunk 5 above)
Comment (meandavejustice) [inline gulpfile.js]: Thanks for bringing this up. I should probably move over to only using gulp.

File: CONTRIBUTING.md (hunk 4 above)
Comment (jaredhirsch) [inline CONTRIBUTING.md]: > I guess travis checks to see that the tests are run.

For sure, but _humans_ need to verify that the tests make sense, are reasonably independent of one another (so that one bug doesn't cause lots of seemingly-unrelated tests to fail), and that the most important or tricky corners have a little extra thorough coverage.

File: idea_town/frontend/static-src/app/views/base-view.js [hunk 11]
```diff
@@ -33,7 +33,7 @@ export default AmpersandView.extend({
 
     // HACK: Slap the same data-l10n-args data on every localized node, because
     // the most common case is they all need the same model data.
-    const nodes = this.el.querySelectorAll('[data-l10n-id]');
+    const nodes = this.queryAll('[data-l10n-id]');
```
Comment (jaredhirsch) [inline idea_town/frontend/static-src/app/views/base-view.js:36]: It's fine if you want to include this bug fix in this PR, but it needs to be in a separate commit.

There are tons of ways to do this, but here's one way I'd do it:

``` sh
# pull this file out of the main commit
git reset HEAD~1 idea_town/frontend/static-src/app/views/base-view.js
# amend the main commit
git commit --amend --no-edit
# create a new commit
git add idea_town/frontend/static-src/app/views/base-view.js
git commit -m "Use an array, not a NodeList, to allow for-of iteration of l10n nodes"
```

From there, you can `git rebase -i HEAD~2` and swap the order of the two commit lines to put the bug fix earlier in history than the main testing commit.

Comment (meandavejustice) [inline idea_town/frontend/static-src/app/views/base-view.js:36]: Thanks @6a68 I'll take care of that :)

**Comment by jaredhirsch** [comment]
@meandavejustice OK, this has been open a really long time, so let's get it landed. Things I think are essential and need to be addressed:

Bigger things:
- Because this PR has been dragging on, if the view test bug can't be fixed in an hour (the `can't find variable: Symbol` bug), just pull those tests out of this PR, stash them in a branch somewhere, and file a followup bug to add view tests. Timebox this aggressively: there be dragons in generated code debugging :dragon_face: 
- The test runner has some problems currently: (1) test runner shouldn't bail if a test throws, and (2) tests currently don't call teardown code if a test fails/throws. To fix this, add graceful exception handling, and separate invocation of setup/test/teardown methods (or beforeEach/afterEach, whatever). The [gist](https://gist.github.com/6a68/c262519b2c1fee3fd251) I mentioned in IRC could work, or use whatever solution you like, so long as we solve the two problems.
- Once the test runner has before/after capability, refactor the tests to pull before/after code out into separate methods (figure out whether you're keeping the view tests before you bother refactoring them). Organize the code however you'd like, just be sure that:
  - shared setup isn't duplicated
  - teardown is in a separate method (to guard against test exceptions), and restores fixture state regardless of whether the test passed, failed, or threw.

Smaller things:
- Get rid of the weird unicode bits (`🖳`  and `␄`) at the start and end of the test names. The test runner output would look fine without them, and this could be confusing to less-experienced contributors.
- Test names should be consistently uppercased or downcased, pick any one. Try to make the format of the test name as consistent and predictable as possible, to make it easy for contributors to add or split tests.
- If you do keep the view tests in this PR, rather than test for the existence of innerHTML, it would be better to test for the presence of a specific node that's inserted when the template has rendered correctly. Any attribute is fine, l10n ids or classes or ids, just test that the right template has rendered.
- Tests are easier to trust when they are obvious, so I'd favor obviousness over terseness. Try to do one thing per line (within reason). In particular, please separate the test code from the assert code:

``` js
// this is really terse
t.ok(new View({headerScroll:false}).render().el);

// this version is really simple: easy to scan, separates the test
// from the assertion, and is more friendly to a novice contributor
const view = new View({headerScroll: false});
view.render();
const someEl = view.query('#some-el-in-template');

t.ok(someEl);
```

Please ping me if you have questions or get stuck on any of these items. Once this list is addressed, I'll probably do one quick nitpicky review, aimed at minor line-level syntactic things, then we'll land this. :+1:

**Comment by jaredhirsch** [comment]
Ah, one other thing I missed: when everything else is done, please expand the commit message, as Les mentioned [above](https://github.com/mozilla/idea-town/pull/354#issuecomment-171372474). Happy to offer feedback/advice on this step too. There are no stupid questions :-)

File: idea_town/frontend/static-src/app/views/base-view.js (hunk 11 above)
Comment (meandavejustice) [inline idea_town/frontend/static-src/app/views/base-view.js:36]: https://github.com/meandavejustice/idea-town/commit/454e232792a4cf8dbb07bc87e89338ea47be1519

**Comment by meandavejustice** [comment]
@6a68 I believe I have addressed all of your concerns. Thanks for the feedback.

**Comment by meandavejustice** [comment]
@6a68 or @lmorchard r?

**Comment by lmorchard** [comment]
lgtm, let's get this thing merged
